lib/conrad/collector.rb

```diff
@@ -0,0 +1,168 @@
+    attr_accessor :event_metadata
+
+    # The events stored in the collector
+    attr_reader :events
+
+    # ProcessorStack used on each event added
+    attr_reader :processors
+
+    # Formatter used to generate sendable format for an Event
+    attr_accessor :formatter
+
+    # Emitter used to send out events
+    attr_accessor :emitter
+
+    # Boolean indicating if events should be sent as a batch or individually by
+    #   default for each Collector instance
+    attr_accessor :emit_as_batch
+    alias emit_as_batch? emit_as_batch
+
+    # @param processors [Array<#call>] set of processors to run. Defaults to
+    #   processors as configured for the class.
+    # @param formatter [#call] Formatter to use. Defaults to
+    #   formatter as configured for the class.
+    # @param emitter [#call] emitter to send events. Defaults to
+    #   emitter as configured for the class.
+    # @param emit_as_batch [Boolean] indicates how to send events. Defaults to
+    #   value configured for class.
+    def initialize(
+      processors: self.class.default_processors,
+      formatter: self.class.default_formatter,
+      emitter: self.class.default_emitter,
+      emit_as_batch: self.class.default_emit_as_batch
+    )
+      @events = []
+      @event_metadata = {}
+      @processors = processors
+      @processor_stack = Conrad::ProcessorStack.new(processors)
+      @formatter = formatter
+      @emitter = emitter
+      @emit_as_batch = emit_as_batch
+    end
+
+    # Adds an event to the Collector to be audited at a later time. The
+    # current collector's event_metadata is added to the event, it is processed,
+    # then it is formatter before being added to the set of events.
+    # If `:halt_conrad_processing` is thrown during the event processing, then
+    # the event will not be added to the collection.
+    #
+    # @param event [Hash]
+    #
+    # @raise [ForbiddenKey] when a key is neither a Symbol nor a String
+    def add_event(event)
+      processed_event = processor_stack.call(event.merge(event_metadata))
+
+      return unless processed_event
+
+      validate_event_keys(processed_event)
+
+      events << formatter.call(processed_event)
+    end
+
+    # Records the events currently in the collection then clears the state of
+    # the Collector by emptying the events stack and clearing out the metadata.
+    def record_events
+      if emit_as_batch?
+        record_events_as_batch
+      else
+        record_individual_events
+      end
+
+      reset_state
+    end
+
+    # Attribute writer for changing the processors for an instance of a
+    # Collector
+    def processors=(processors)
+      @processor_stack = Conrad::ProcessorStack.new(processors)
+      @processors = processors
+    end
+
+    private
+
+    attr_reader :processor_stack
+
+    def validate_event_keys(event)
+      event.each_key do |key|
+        raise ForbiddenKey, key unless key.is_a?(Symbol) || key.is_a?(String)
```

Comment:
would it be helpful to have more info about the error in here?

Reply:
What do you mean by that? The error itself is documented in the `errors` file.

Reply:
By that I mean that I was too lazy to look in the errors file. haha
That I meant initially is something like a verbose description along the lines of, `"The key '#{key}' is invalid. Must be a symbol or a string."`
But since it's a `Conrad::ForbiddenKey` error, it's really not necessary.